sdk/core/azure-core/inc/azure/core/hash.hpp

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file
+ * @brief Utility functions to help compute the hash value for the input binary data, using
+ * algorithms such as MD5.
+ */
+
+#pragma once
+
+#include <cstdint>
+#include <stdexcept>
+#include <vector>
+
+namespace Azure { namespace Core {
+
+  /**
+   * @brief Represents the base class for hash algorithms which map binary data of an arbitrary
+   * length to small binary data of a fixed length.
+   */
+  class Hash {
+  private:
+    /**
+     * @brief Used to append partial binary input data to compute the hash in a streaming fashion.
+     * @remark Once all the data has been added, call #Final() to get the computed hash value.
+     * @param data The pointer to the current block of binary data that is used for hash
+     * calculation.
+     * @param length The size of the data provided.
+     */
+    virtual void OnAppend(const uint8_t* data, std::size_t length) = 0;
```

Comment:
Append seems a better name.  Why is it prefixed with 'On'

Reply:
Following existing `OnRead`/`Read` pattern from APIs like BodyStream, where the `OnX` method is private virtual, meant for derived  classes to implement, and the `X` method is public:
https://github.com/Azure/azure-sdk-for-cpp/blob/568d4bcc4e04b34478d1bb634c672ef2e6e8c67f/sdk/core/azure-core/inc/azure/core/http/body_stream.hpp#L52-L87